### PR TITLE
Implement capacity, insert and insertBuffer

### PR DESCRIPTION
**Pull request by darkdrag00nv2**
As part of this PR, I've implemented three methods in StableBuffer:
- capacity
- insert
- insertBuffer

Also added tests for them.

### CODE REVIEW

File: src/StableBuffer.mo [hunk 1]
```diff
@@ -170,4 +189,138 @@ module {
...
+  ///
+  /// Amortized Space: O(1), Worst Case Space: O(size1 + size2)
+  public func insertBuffer<X>(buffer: StableBuffer<X>, index : Nat, buffer2 : StableBuffer<X>) {
+    if (index > buffer.elems.size()) {
+      Prim.trap "Buffer index out of bounds in insertBuffer"
+    };
+    // exit early if the second buffer is empty.
+    if (buffer2.count == 0) {
+      return
+    };
```
Comment (ByronBecker) [inline src/StableBuffer.mo:282]: Nice 👍

File: src/StableBuffer.mo [hunk 2]
```diff
@@ -170,4 +189,138 @@ module {
...
+    let size2 = buffer2.count;
+    let capacity = buffer.elems.size();
+
+    // copy elements to new array and shift over in one pass
+    if (buffer.count + size2 > capacity) {
+      let placeholder = if (buffer.count > 0) {
+        buffer.elems[0]
+      } else {
+        buffer2.elems[0]
+      };
```
Comment (ByronBecker) [inline src/StableBuffer.mo:293]: Good catch!

File: test/StableBufferTest.mo [hunk 3]
```diff
@@ -113,5 +119,292 @@ do {
...
+        ),
+        test(
+          "elements",
+          B.toArray(buffer),
+          M.equals(T.array<Nat>(T.natTestable, [10, 11, 12, 13, 14, 15, 0, 1, 2, 3, 4, 5]))
+        )
+      ]
+    )
+  );
+};
```
Comment (ByronBecker) [inline test/StableBufferTest.mo:223]: Thanks for including these matchers tests! They are much more hardy than just the asserts.

One suggestion (if you have the time) is that you can actually stack multiple suites together like this.

```
run(
  suite(
    "insertBuffer at end",
      [
        ...tests
      ]
  ),
  suite("insertBuffer with capacity change",
    [
       ...tests
    ]
  ),
  ...
);
```

This is the same pattern that I've been using for the rest of my libraries. The reason this library still has assert tests is that it was one of the first libraries I wrote and originally copied over from base (base used to be all assert tests as well 😅)

**Review by ByronBecker — APPROVED** [review]
Logic looks great, one small suggestion - not a huge deal if you don't have the time to complete in the next week or so, I'll just merge in a week or so from now if there hasn't been any changes made

File: test/StableBufferTest.mo (hunk 3 above)
Comment (darkdrag00nv2) [reply test/StableBufferTest.mo:223]: Sure, I'll get to it in 1-2 days. I should get some time then.

**Pull request closed without merging**